compiler/noirc_evaluator/src/ssa/ir/instruction.rs

```diff
@@ -744,6 +739,65 @@ impl Instruction {
     }
 }
 
+/// Given a chain of operations like:
+/// v1 = array_set [10, 11, 12], index 1, value: 5
+/// v2 = array_set v1, index 2, value: 6
+/// v3 = array_set v2, index 2, value: 7
+/// v4 = array_get v3, index 1
+///
+/// We want to optimize `v4` to `10`. To do this we need to follow the array value
+/// through several array sets. For each array set:
+/// - If the index is non-constant we fail the optimization since any index may be changed
+/// - If the index is constant and is our target index, we conservatively fail the optimization
+///   in case the array_set is disabled from a previous `enable_side_effects_if` and the array get
+///   was not.
+/// - Otherwise, we check the array value of the array set.
+///   - If the array value is constant, we use that array.
+///   - If the array value is from a previous array-set, we recur.
+fn try_optimize_array_get_from_previous_set(
+    dfg: &mut DataFlowGraph,
+    mut array_id: Id<Value>,
+    target_index: FieldElement,
+) -> SimplifyResult {
+    let mut elements = None;
+
+    // Arbitrary number of maximum tries just to prevent this optimization from taking too long.
+    let max_tries = 5;
+    for _ in 0..max_tries {
+        match &dfg[array_id] {
+            Value::Instruction { instruction, .. } => {
+                match &dfg[*instruction] {
+                    Instruction::ArraySet { array, index, value, .. } => {
+                        if let Some(constant) = dfg.get_numeric_constant(*index) {
+                            if constant == target_index {
+                                return SimplifyResult::SimplifiedTo(*value);
+                            }
+
+                            array_id = *array; // recur
+                        } else {
+                            return SimplifyResult::None;
```

Comment:
Haven't thought about whether this is desirable but we could also simplify this so that we perform the read on the earliest suitable array, i.e.

```
v1 = array_set [10, 11, 12], index v0, value: 5
v2 = array_set v1, index 2, value: 6
v3 = array_set v2, index 2, value: 7
v4 = array_get v3, index 1

// goes to

v1 = array_set [10, 11, 12], index v0, value: 5
v2 = array_set v1, index 2, value: 6
v3 = array_set v2, index 2, value: 7
v4 = array_get v1, index 1
```

Reply:
This may end up being counterproductive in some situations as can prevent mutation of `v1`.

Reply:
It could possibly be useful in combination with instruction de-duplication.

I think I'll merge this PR then make a separate PR to test this comment.

Reply:
Yep, agreed on tackling this separately.